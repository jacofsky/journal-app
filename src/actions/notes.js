import { addDoc, collection, deleteDoc, doc, updateDoc } from "firebase/firestore";
import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

// react-journal

export const startNewNote = () => {
    return async (dispatch, getState) => {
        
        const uid = getState().auth.uid;
        console.log(uid) 

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await addDoc(collection(db, `${uid}`, "journal/notes"), newNote)

        console.log("Document written with ID: ", doc);

        dispatch(activeNote(doc.id, newNote))
        dispatch(addNetNote(doc.id, newNote))


    
    }
}

export const activeNote = (id, note) => ({

    type: types.notesActive,
    payload: {
        id, 
        ...note
    }

}) 

export const addNetNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {
        const uid = getState().auth.uid;

        if(!note.url){
            delete note.url
        }

        const noteToFirestore = {...note}
        delete noteToFirestore.id;

        await updateDoc(doc(db, `${uid}/journal/notes/${note.id}`), noteToFirestore)

        dispatch(refreshNote(note.id, noteToFirestore))
        Swal.fire({
            icon: 'success',
            title: 'Saved',
            text: note.title,
            confirmButtonColor: '#5cc5bc'
        })
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState) => {

        const {active:activeNote} = getState().notes;

        
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading()
            }
        })
        
        const imgUrl = await fileUpload(file)  
        activeNote.url = imgUrl;
        
        dispatch(startSaveNote(activeNote)) 
        
        
        Swal.close()  
        
        
    }
}

export const startDeleating = (id) => {
    return async(dispatch, getState) => {

        const uid = getState().auth.uid;

        await deleteDoc(doc(db, `${uid}/journal/notes/${id}`))

        dispatch(deleteNote(id))
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const logoutCleaning = () => ({
    type: types.notesLogoutCleaning
})