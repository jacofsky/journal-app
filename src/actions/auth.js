import Swal from 'sweetalert2'

import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import {types} from '../types/types'
import { finishLoading, startLoading } from './ui';
import { logoutCleaning } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading())
        const auth = getAuth()
        signInWithEmailAndPassword(auth, email, password)
            .then(({user}) => {      
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(e =>{ 
                console.log(e) 
                dispatch(finishLoading())
                Swal.fire({
                    icon: 'error',
                    title: 'Oops...',
                    text: 'Check your email or password',
                    confirmButtonColor: '#5cc5bc'
                  })
            })
    
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        const auth = getAuth()
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({user}) => {
                await updateProfile(auth.currentUser, {displayName: name})
                
                dispatch(login(user.uid, user.displayName))
            })
            .catch (e => Swal.fire({
                icon: 'error',
                title: 'Oops...',
                text: 'The email has already an account',
                confirmButtonColor: '#5cc5bc'
              }))
    
    }
}

export const startGoogleLogin = () => {
    
    return (dispatch) => {
        
        const auth = getAuth()
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName))
            })
            
    }

}

export const login = (uid, displayName) => ({
        
    type: types.login,
    payload: {
        uid,
        displayName
    }
    
})

export const startLogout = () => {
    return async(dispatch) => {
        const auth = getAuth()
        await auth.signOut()

        dispatch(logout())
        dispatch(logoutCleaning())
    }
}

export const logout = () => ({
    
    type: types.logout

})