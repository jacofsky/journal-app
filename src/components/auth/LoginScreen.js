import React, { useEffect } from 'react'
import {useDispatch, useSelector} from 'react-redux'

import 'bootstrap-icons/font/bootstrap-icons.css';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';


const LoginScreen = () => {

    const dispatch = useDispatch()
    const {loading, msgError} = useSelector(state => state.ui)

    useEffect(() => {
        dispatch(removeError())
    }, [])

    const {formValues, handleInputChange } = useForm({
        email: 'Pablis@gmail.com',
        password: '123456'
    })

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault()
        if(isFormValid()) {
            dispatch(startLoginEmailPassword(email, password))
        }

    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin())
        
    }

    const isFormValid = () => {
        
        

        if (!validator.isEmail(email)){
            dispatch(setError('Email is invalid'))


            return false 
        } else if (validator.isEmpty(password)) {
            dispatch(setError('Password is required'))

            return false
        }

        dispatch(removeError())
        return true
    }

    return (
        <div className='animate__animated animate__fadeIn animate__faster'>


            <h3 className='auth__tittle'>Login</h3>

            {
                msgError 
                &&
                <div className="auth__alert-error my-2">
                    {msgError}
                </div>
            }

            <form onSubmit={handleLogin}>
                <div className="row">
                    <div>
                        <input className='w-100 mb-2 auth__input' type="text" placeholder='Email' name='email' value={email} onChange={handleInputChange} />
                        <input className='w-100 mt-2 auth__input' type="password" placeholder='Password' name='password' value={password} onChange={handleInputChange} />
                    </div>
                    <div className='justify-content-center d-flex mt-3'>
                        <button className="auth__button btn w-100 justify-content-center d-flex" type="submit" disabled={loading}>
                            {
                            
                            loading 
                            ? 
                            <div className="spinner-border" role="status">
                                <span className="visually-hidden">Loading...</span>
                            </div> 
                            : 
                            <p>Login <i className="bi bi-arrow-bar-right"></i></p> 
                            
                            } 
                        </button>
                    </div>
                    
                </div>
            </form>
                
                
                
                
            <div className='my-4 d-flex justify-content-center align-items-center'>
                <hr className='col-5'/>
                <p className='text-center col-2'>or</p>   
                <hr className='col-5'/>
            </div>


            <div className="d-flex justify-content-center pb-4">
                <div className="text-center">
                    <button onClick={handleGoogleLogin} className='btn btn-outline-secondary'><i className="bi bi-google auth-media"></i></button>
                </div>
            </div>

            <Link to="/auth/register" className='link'>
                Create new account
            </Link>
                
        </div>
    )
}

export default LoginScreen
