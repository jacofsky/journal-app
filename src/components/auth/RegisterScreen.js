import React, { useEffect } from 'react'
import {useDispatch, useSelector} from 'react-redux'

import { Link } from 'react-router-dom';
import validator from 'validator';


import { useForm } from '../../hooks/useForm';
import { setError, removeError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';


const RegisterScreen = () => {

    const dispatch = useDispatch()
    const {msgError} = useSelector( state => state.ui )

    useEffect(() => {
        dispatch(removeError())
    }, [])

    const {formValues, handleInputChange} = useForm({
        name: 'Pablin',
        email: 'Pablis@gmail.com',
        password: '123456',
        password2: '123456'
    })

    const {name, email, password, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault()

        if(isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name))
        }
    }

    const isFormValid = () => {
        
        if (validator.isEmpty(name)) {
            dispatch(setError('Name is required'))

            return false
        } else if (!validator.isEmail(email)){
            dispatch(setError('Email is invalid'))


            return false 
        } else if (password !== password2 || password.length < 5 || password2.length < 5) {
            dispatch(setError('Password shold be at least 6 caracters and match each other'))

            return false 
        }

        dispatch(removeError())
        return true
    }

    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className='auth__tittle'>Register</h3>
            
            {
                msgError 
                &&
                <div className="auth__alert-error my-2">
                    {msgError}
                </div>
            }
            

            <form onSubmit={handleRegister}>
                <div className="row mb-3">
                    <div>
                        <input className='w-100 mb-2 auth__input' type="text" placeholder='Name' name='name' onChange={handleInputChange} value={name} />
                        <input className='w-100 mb-2 auth__input' type="text" placeholder='Email' name='email' onChange={handleInputChange} value={email} />
                        <input className='w-100 mt-2 auth__input' type="password" placeholder='Password' name='password' onChange={handleInputChange} value={password} />
                        <input className='w-100 mt-2 auth__input' type="password" placeholder='Confirm password' name='password2' onChange={handleInputChange} value={password2} />
                    </div>
                    <div className='justify-content-center d-flex mt-3'>
                        <button className="auth__button btn w-100" type="submit">Register <i className="bi bi-arrow-bar-right"></i></button>
                    </div>
                    
                </div>

                <Link to="/auth/login" className='link'>
                    Already registered?
                </Link>
                
            </form>
        </div>
    )
}

export default RegisterScreen
