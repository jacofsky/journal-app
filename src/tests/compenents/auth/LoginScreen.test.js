import React from "react"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router-dom"
import LoginScreen from "../../../components/auth/LoginScreen"
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount } from "enzyme"
import { startGoogleLogin, startLoginEmailPassword } from "../../../actions/auth"

jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {

    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
)

describe('Tests in LoginScreen', () => {


    beforeEach(() => {
        store = mockStore(initState)
        jest.clearAllMocks()
    })
    


    test('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should startGoogleLogin', () => { 
        
        wrapper.find('.btn-outline-secondary').prop('onClick')()

        expect(startGoogleLogin).toHaveBeenCalled()

    })

    test('should handleLogin', () => { 
        
        wrapper.find('form').prop('onSubmit')({preventDefault(){}})

        expect(startLoginEmailPassword).toBeCalledWith('Pablis@gmail.com','123456')

    })


})