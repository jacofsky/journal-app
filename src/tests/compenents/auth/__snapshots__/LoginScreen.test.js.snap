// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in LoginScreen should match snapshot 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__tittle"
  >
    Login
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div>
        <input
          className="w-100 mb-2 auth__input"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value="Pablis@gmail.com"
        />
        <input
          className="w-100 mt-2 auth__input"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value="123456"
        />
      </div>
      <div
        className="justify-content-center d-flex mt-3"
      >
        <button
          className="auth__button btn w-100 justify-content-center d-flex"
          disabled={false}
          type="submit"
        >
          <p>
            Login 
            <i
              className="bi bi-arrow-bar-right"
            />
          </p>
        </button>
      </div>
    </div>
  </form>
  <div
    className="my-4 d-flex justify-content-center align-items-center"
  >
    <hr
      className="col-5"
    />
    <p
      className="text-center col-2"
    >
      or
    </p>
    <hr
      className="col-5"
    />
  </div>
  <div
    className="d-flex justify-content-center pb-4"
  >
    <div
      className="text-center"
    >
      <button
        className="btn btn-outline-secondary"
        onClick={[Function]}
      >
        <i
          className="bi bi-google auth-media"
        />
      </button>
    </div>
  </div>
  <a
    className="link"
    href="/auth/register"
    onClick={[Function]}
  >
    Create new account
  </a>
</div>
`;
